src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+begin
+  ext x,
+  rw [inv_apply_eq_inv],
+  rcases hχ x with h₀ | h₁ | h₂,
+  { rw [h₀, ring.inverse_zero], },
+  { rw [h₁, ring.inverse_one], },
+  { rw [h₂, (by norm_num : (-1 : R') = (-1 : R'ˣ)), ring.inverse_unit (-1 : R'ˣ)],
+    refl, },
+end
+
+/-- The square of a quadratic character is the trivial character. -/
+lemma quad_char_sq_eq_one {χ : mul_char R R'} (hχ : χ.is_quadratic) : χ ^ 2 = 1 :=
+begin
+  rw [pow_two],
+  nth_rewrite 0 ← quadratic_char_inv hχ,
+  rw [mul_left_inv],
+end
+
+/-- The `p`th power of a quadratic character is itself, when `p` is the (prime) characteristic
+of the target ring. -/
+lemma quad_char_pow_char {χ : mul_char R R'} (hχ : χ.is_quadratic)
+ (p : ℕ) [hp : fact p.prime] [char_p R' p] :
+  χ ^ p = χ :=
+begin
+  ext x,
+  rw [pow_apply_coe],
+  rcases hχ x with (hx | hx | hx); rw hx,
+  { rw [zero_pow (fact.out p.prime).pos], },
+  { rw [one_pow], },
+  { exact char_p.neg_one_pow_char R' p, },
+end
+
+/-- The `n`th power of a quadratic character is itself, when `n` is odd. -/
+lemma quad_char_pow_odd {χ : mul_char R R'} (hχ : χ.is_quadratic) {n : ℕ} (hn : n % 2 = 1) :
+  χ ^ n = χ :=
+by rw [← nat.div_add_mod n 2, hn,
+       pow_add, pow_mul, quad_char_sq_eq_one hχ, one_pow, pow_one, mul_char.one_mul]
+
+open_locale big_operators
+
+/-- The sum over all values of a nontrivial multiplicative character on a finite ring is zero
+(when the target is a domain). -/
+lemma sum_eq_zero_of_is_nontrivial [fintype R] [is_domain R'] {χ : mul_char R R'}
+ (hχ : χ.is_nontrivial) :
+  ∑ a, χ a = 0 :=
+begin
+  rcases hχ with ⟨b, hb⟩,
+  have h₁ : ∑ a, χ (b * a) = ∑ a, χ a :=
+    fintype.sum_bijective _ (units.mul_left_bijective b) _ _ (λ x, rfl),
+  simp only [map_mul] at h₁,
+  rw [← finset.mul_sum] at h₁,
+  exact eq_zero_of_mul_eq_self_left hb h₁,
+end
+
+/-- The sum over all values of the trivial multiplicative character on a finite ring is
+the cardinality of its unit group. -/
+lemma sum_eq_card_units_of_is_trivial [fintype R] [decidable_eq R] :
+  ∑ a, (1 : mul_char R R') a = fintype.card Rˣ :=
+begin
+  have h₁ : ∀ a : R, (1 : mul_char R R') a = ite (is_unit a) 1 0 :=
+  by { intro a, split_ifs, { exact one_apply_coe h.unit }, { exact map_nonunit _ h } },
+  simp_rw [h₁],
+  have h₂ := @finset.sum_filter R' R finset.univ _ is_unit _ 1,
+  simp only [pi.one_apply] at h₂,
+  have h₃ := map_sum (algebra_map ℕ R') 1 (finset.filter (is_unit : R → Prop) finset.univ),
+  simp only [pi.one_apply] at h₃,
+  rw [← finset.card_eq_sum_ones, eq_nat_cast, eq_nat_cast, nat.cast_one] at h₃,
+  have h₄ : finset.filter (is_unit : R → Prop) finset.univ
+             = finset.map ⟨(coe : Rˣ → R), units.ext⟩ finset.univ :=
+  begin
+    ext a,
+    simp only [finset.mem_filter, finset.mem_univ, true_and, finset.mem_map,
+               function.embedding.coe_fn_mk, exists_true_left],
+    refl,
+  end,
+  rw [← h₂, ← h₃, h₄, finset.card_map, fintype.card],
+end
```

Comment:
A neat way to structure these big calculational proofs is the `calc` block.
```suggestion
begin
  calc ∑ a, (1 : mul_char R R') a
      = ∑ a : R, if is_unit a then 1 else 0 : finset.sum_congr rfl (λ a _, _)
  ... = ((finset.univ : finset R).filter is_unit).card : finset.sum_boole
  ... = (finset.univ.map (⟨(coe : Rˣ → R), units.ext⟩)).card : _
  ... = fintype.card Rˣ : congr_arg _ (finset.card_map _),
  { split_ifs with h h,
    { exact one_apply_coe h.unit },
    { exact map_nonunit _ h } },
  { congr,
    ext a,
    simp only [finset.mem_filter, finset.mem_univ, true_and, finset.mem_map,
               function.embedding.coe_fn_mk, exists_true_left, is_unit] }
end
```

Reply:
Yes, this looks a bit better. I was struggling a bit with this proof, fiddling with the sums and finsets...